module/errorhandler/erroraware.go

```diff
@@ -0,0 +1,165 @@
+package errorHandler
+
+import (
+	"context"
+	"log"
+	"runtime"
+
+	"github.com/onflow/flow-go/module"
+)
+
+// IrrecoverableSignaler sends the error out
+// little variability on what happens => struct
+type IrrecoverableSignaler struct {
+	irrecoverableErrors chan<- error
+}
+
+// ThrowIrrecoverable is a narrow drop-in replacement for panic, log.Fatal, log.Panic, etc
+// anywhere there's something connected to the irrecoverable Error channel
+func (e *IrrecoverableSignaler) ThrowIrrecoverable(err error) {
+	e.irrecoverableErrors <- err
+	runtime.Goexit()
+}
+
+// IrrecoverableHandler reacts to the error
+// lots of variability in how the reaction could go:
+// - restart the component (in production) after cleanup
+// - panic (in canary / benchmark)
+// - log in various Error channels and / or send telemetry ...
+// does not do the plumbing of connecting to the channel
+type IrrecoverableHandler interface {
+	// OnIrrecoverable handles the error, and
+	// executes andThen.
+	// despite the restart use case, andThen
+	// is actually a general instance of a continuation
+	OnIrrecoverable(err error, andThen func())
```

Comment:
When a handler has multiple `IrrecoverableSignaler` enabled child components, it will need to be able to differentiate which one threw the error. Maybe we need to pass back a reference to the component here?

Reply:
> When a handler has multiple IrrecoverableSignaler enabled child components

Ideally you wouldn't: the key to the hierarchical model enforced by passing the `IrrecoverableSignaler` through a context is that every sub-component hides the parent's `IrrecoverableSignaler` and replaces it with its own, handled by its parent.

So: one `IrrecoverableSignaler` per component, handled by your *direct* parent.

> potentially require a cleanup method on the component that can be called by the handler

AFAICT that's the role of `cancel()`

Reply:
Yeah, @huitseeker overall I like the `IrrecoverableSignaler` and `IrrecoverableSignalerContext`, but I'm not a fan of defining `IrrecoverableHandler` as an interface rather than a function. One reason is the one peter mentioned above, but another reason is because it makes the error handling logic less clear since it is farther away from the logic which runs the component.

If the error handling logic is defined as a function, we can see exactly how error handling is performed at the exact same place in the code where `RunComponent` is called, making the code very easily understandable.

On the otherhand, with an interface the actual implementation will live elsewhere, possible even in a different file but at the very least in a separate struct. Furthermore, you might need multiple different implementations for different components, and this quickly gets messy.

The interface only has a single method, I say we just get rid of the interface and use a function instead.

Reply:
> AFAICT that's the role of cancel()

Do you mean the cancel in RunComponent [here](https://github.com/onflow/flow-go/pull/1308/files#diff-2e3ca680b4c761f36416e96719a5e92052aad1633269f7bf4c1a40189664bc69R151)?
At that point, the component has already called `Goexit()` and its goroutine has exited, so there shouldn't be anything listening for done events on the other end of that cancel.

Reply:
At the end of the day, the model I care about is this: a component starts k sub-components. In doing so, it manages k channels which pass the irrecoverables of those k sub-components and manages them accordingly. If any one of these sub-components spawns sub-components, the same pattern applies for irrecoverables.

I'm OK with using a function here, just beware that it may make reusing handler implementations a tad less easy.

> If the error handling logic is defined as a function, we can see exactly how error handling is performed at the exact same place in the code where RunComponent is called, making the code very easily understandable.

A function is reusable, but I think that precise point will make us have "irrecoverable policy" functions that are defined away from their component.

Reply:
> At the end of the day, the model I care about is this: a component starts k sub-components. In doing so, it manages k channels which pass the irrecoverables of those k sub-components and manages them accordingly. If any one of these sub-components spawns sub-components, the same pattern applies for irrecoverables.

Yes, we are on the same page about this, and in general I like the suggestions made here. Only getting into the details because we are now at the stage of actually writing the code that will be merged to the codebase.

I think functions give us more flexibility. If we need one that is reusable, we can define it separately. If it doesn't need to be reusable, then we can define the function inline, giving us better code clarity.